drivers: video: Add support for arducam mega

ArduCAM opened the pull request:
The Arducam mega is an low power, rolling shutter camera, support connect one or more cameras any Microcontroller. It provides high-quality image capture and processing capabilities, making it highly suitable for various application fields, including machine vision, image recognition, and robotics, among others.

This commit adds arducam mega driver and bindings file.

marcusfolkesson commented on drivers/video/Kconfig.arducam_mega:
```diff
@@ -0,0 +1,12 @@
+	bool "Arducam Mega CMOS digital image sensor"
+	select SPI
+	depends on DT_HAS_ARDUCAM_MEGA_ENABLED
+	default y
```
default y ?

ArduCAM replied:
@marcusfolkesson 
Thank you for your detail check.
Yes, Here we refer to this [video drivers](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/video/). we find all of those driver's configuration are defaut y.

bjarki-trackunit commented:
Hi, the zephyr project uses the Apache 2.0 license [See](https://docs.zephyrproject.org/latest/contribute/guidelines.html#licensing), with exceptions for code imported from other projects, like this [example](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/testsuite/include/zephyr/fff.h) but it seems this PR is written specifically for Zephyr, could you update the PR to use the Apache 2.0 license?

bjarki-trackunit commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1064 @@
+#include <zephyr/drivers/video.h>
+#include <zephyr/drivers/spi.h>
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mega_camera);
```
You can simply use 
```
#include <zephyr/logging/log.h>
LOG_MODULE_REGISTER(mega_camera);
```
to achieve the same :)

bjarki-trackunit commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1064 @@
+	uint8_t stream_on;
+};
+
+static struct arducam_mega_info mega_infos[] = {{
```
Make these const to store them in ROM if they are not modified during runtime

bjarki-trackunit commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1064 @@
+		.height_min = (height), .height_max = (height), .width_step = 0, .height_step = 0  \
+	}
+
+static struct video_format_cap fmts[] = {
```
Make const

ArduCAM replied:
@bjarki-trackunit 
Thank you for your detail check. In fact, the arducam mega driver is used for multi types camera instead of one specific sensor, so the content of the fmts will be changed when connect different cameras which mega driver has supported. That's why we define it in static struct video_format_cap.

bjarki-trackunit replied:
ok

bjarki-trackunit commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1064 @@
+
+#define SUPPORT_RESOULTION_NUM 9
+
+static uint8_t support_resoultion[SUPPORT_RESOULTION_NUM] = {
```
Make const

ArduCAM replied:
@bjarki-trackunit 
Thank you for your detail check. This is the same as the [fmts](https://github.com/zephyrproject-rtos/zephyr/pull/66994#discussion_r1436182143)， due to the arducam mega driver supports multi cameras.

bjarki-trackunit replied:
ok

bjarki-andreasen reviewed (changes requested):
Initial impressions are good :)

loicpoulain commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1064 @@
+
+#define DT_DRV_COMPAT arducam_mega
+
+#include <zephyr/drivers/video/arducam_mega.h>
```
Why do you need an header file? as far as I see everything can be in the C file.

ArduCAM replied:
@loicpoulain 
Thank you for your detail check.
Due to our arducam mega driver need some specific controls and other parameters and we need provide them to user.
We have not see them in the video.h file. So we add a arducam_mega.h file. If not, users can't control our mega's full features.

loicpoulain replied:
Ok, but look like some of them, like exposure, brightness, etc, are quite common for camera, so it would be worth to add new generic CId for them, instead of having custom ones for each vendor. For sure that new control values should be in known unit like `us` or `%`..

ArduCAM replied:
@loicpoulain 
Hello,  Thank you for your detail check. Let me understand your means.
For some common camera controls, it is not possible to define custom own CID, Just use the commn CID in the video.h.
For some specific control, we can define new CID to video.h. Right?
Sorry I don't get it about  "For sure that new control values should be in known device-agnostic unit like us or %..".
How to modify it.

ArduCAM replied:
@loicpoulain 
Hello, thanks for your fantastic sugguestions. In fact， in the video-controls.h file there are exist some common  camera controls, and we have use it，we can add some other common controls（such as exposure mode） to video-controls.h and send you a  new PR.
But，we have some specific control just used for arducam mega. so we can't make sure all controls are common.
Do you think so?

loicpoulain replied:
Yes that's fine, you can also have private controls. But it makes sense to make the generic ones common.

ArduCAM replied:
@loicpoulain 
OK, I have modified it.

loicpoulain commented:
Hello, could you please rebase and squash.

ArduCAM commented:
@loicpoulain 
Hi, I have rebase and squash the PR, Does it have  other issues?

ArduCAM commented:
@galak 
Hello,
Do you think  does arducam mega driver need any other changes?

bjarki-trackunit commented:
You still need to use the Apache 2.0 license, and you have a few CI issues to fix before this can be merged :)

fvergaraBTD commented on include/zephyr/drivers/video/arducam_mega.h:
```diff
@@ -0,0 +1,192 @@
+ * @enum MEGA_STAURATION_LEVEL
+ * @brief Configure camera stauration  level
+ */
+enum MEGA_STAURATION_LEVEL {
```
Typo for saturation?

ArduCAM commented:
@bjarki-trackunit 
Thank you very much for your detail check. Have modified them.

ArduCAM commented on dts/bindings/vendor-prefixes.txt:
```diff
@@ -63,6 +63,7 @@ arc	Synopsys, Inc. (formerly ARC International PLC)
 arctic	Arctic Sand
 arcx	arcx Inc. / Archronix Inc.
 arduino	Arduino
+arducam	ARDUCAM TECHNOLOGY CO., LIMITED
```
@kartben 
Thank you for your detail check. Sorry there were some typos in mega driver, we have fixed them.

ArduCAM commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1025 @@
+#define CAM_REG_CONTRAST_CONTROL              0X23
+#define CAM_REG_SATURATION_CONTROL            0X24
+#define CAM_REG_EV_CONTROL                    0X25
+#define CAM_REG_WHILEBALANCE_CONTROL          0X26
```
@kartben  
Thank you for detail check. Have fixed them.

loicpoulain commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -33,11 +33,11 @@ extern "C" {
+#define VIDEO_CTRL_CLASS_MPEG    0x00020000 /**< MPEG-compression controls */
+#define VIDEO_CTRL_CLASS_JPEG    0x00030000 /**< JPEG-compression controls */
+#define VIDEO_CTRL_CLASS_VENDOR  0xFFFF0000 /**< Vendor-specific class controls */
 /**
```
No change above?

loicpoulain replied:
Can you move generic video changes to dedicated commit?

UCTRONICS replied:
OK, I have modified it

ArduCAM replied:
@loicpoulain  OK， I have send a new PR

bjarki-andreasen reviewed (dismissed):
My comments have been addressed :)

kartben commented:
@ArduCAM  just curious if you'd be interested in contributing a code sample as a follow-up pull request? It would be really great to have a sample combining video and display APIs to actually display captured images on an LCD display. A sample saving captured image to file system would also be nice :) Thanks again for this PR, and thanks in advance if you happen to be interested in adding code samples :)

Commit:
drivers: video: add some video controls ID

Add some common camera's control: VIDEO_CID_CAMERA_EXPOSURE_AUTO,
VIDEO_CID_CAMERA_GAIN_AUTO, VIDEO_CID_CAMERA_WHITE_BAL_AUTO,
VIDEO_CID_JPEG_COMPRESSION_QUALITY

Signed-off-by: Lee Jackson <lee.jackson@arducam.com>

ArduCAM commented:
@kartben We have prepare some samples for mega API. PRs for samples will be provided next.

ArduCAM commented:
@kartben @loicpoulain 
We have sended a sample about how to testing arducam mega’s API #67538. Thank you for your checking.

loicpoulain commented:
@ArduCAM thanks for your contribution, I have a remaining comment, why do you need a custom `VIDEO_CID_ARDUCAM_CAPTURE` control that you also use in your sample to do a capture. Normally controls are only for control path not to get data. what prevents you to do such capture with the generic framework, e.g. following:

- video_enqueue()                /* Enqueue one buffer for capture */
- video_stream_start()          /* Start capture */
- video_dequeue()                /* Retrieve capture image */
- video_stream_stop()

Also it would be good if you could verify your driver works with the generic video sample (`samples/subsys/video/capture)/`).

Other than that it looks good to me.

ArduCAM commented:
@loicpoulain 
Thank you for your detail check. 
Yes,we have tested the capture demo using arducam mega driver and it can works normal. But due to the memory's limitation of some MCU platform such as RP2040 or STM32，we can't get full resolution image size if using video_enqueue(). In order to get full resolution such as 2592x1944， we have to split the image datas to multi block, which is not compatible with the frame buffer framework. About this, do you have some sugguestions?
In order to get the full resolution image and don't modify the main framework, we adapt the pushing ways.

loicpoulain commented:
> @loicpoulain Thank you for your detail check. Yes,we have tested the capture demo using arducam mega driver and it can works normal. But due to the memory's limitation of some MCU platform such as RP2040 or STM32，we can't get full resolution image size if using video_enqueue(). In order to get full resolution such as 2592x1944， we have to split the image datas to multi block, which is not compatible with the frame buffer framework. About this, do you have some sugguestions? In order to get the full resolution image and don't modify the main framework, we adapt the pushing ways.

Ok, it's a good reason, but It would be perfectly fine to adapt the framework for 'fragmented' frames. Maybe the simpler would be to add a `flags` field to the video_buffer struct, and define a `VIDEO_BUF_MORE_DATA` or `VIDEO_BUF_FRAG` flag so that you know frame is fragmented. That would be a generic way of doing this, assuming your driver can know and indicate it's a fragmented frame.

For a frame capture, that would give something like:
```
 video_enqueue(small_vbuf)
 video_stream_start()
do {
    video_dequeue(small_vbuf) 
    push_data(small_vbuf)
    if (small_vbuf->flags & VIDEO_BUF_MORE) {
        video_enqueue(small_vbuf)
        continue
    }
    break  /* frame complete */
} while (1)
video_stream_stop()
```

Commit:
driver: video: add the fields of the video_buffer structure

Adding frame length and fragmented frame status to indicate certain
information about the frame when the video buffer is smaller than
the frame length.

Signed-off-by: Lee Jackson <lee.jackson@arducam.com>

Commit:
drivers: video: add support for arducam mega

The Arducam mega is an low power, rolling shutter camera,
support connect one or more cameras any Microcontroller.
It provides high-quality image capture and processing
capabilities, making it highly suitable for various
application fields, including machine vision, image
recognition, and robotics, among others.

Signed-off-by: Lee Jackson <lee.jackson@arducam.com>

Commit:
samples: subsys: video: add arducam mega camera full-featured sample

This example is a full-featured sample for the Arducam Mega camera
driver. By using serial communication with a PC software, it can
capture images, display them, and also control camera settings
such as exposure and gain.

Signed-off-by: Lee Jackson <lee.jackson@arducam.com>

ArduCAM commented:
@loicpoulain 
Thank you very much. We have modified the sample and send  new commit. Please check it again.

ArduCAM commented:
@nashif @loicpoulain 
Hi，Does the arducam mega driver and sample need any other modification?

loicpoulain commented:
> @nashif @loicpoulain Hi，Does the arducam mega driver and sample need any other modification?

Thanks for the changes, but the user_data pointer still a arducam specific API we could get avoid. If you introduce to video generic flags (VIDEO_FRAME_FRAGMENTED and VIDEO_FRAME_DONE) you should be able to perform the same and detect when a frame start and finish. [f=frag][f=frag][f=frag|done].

ArduCAM commented:
@loicpoulain 
Your suggestion is very good. We have converted the API to a universal one, but we currently have a problem. We want each fragment frame to have a frame length. The current method that comes to mind is to store it in a structure.
such as
     struct video_frame_info {
         int flags
         uint32 frame_length;
     }
Add the structure to the video_buffer in the form of a pointer. Do you have any suggestions for us to modify this? Your suggestions are very good. I will consider converting the API to a universal one, but we currently have a problem and hope that every fragment Frames all have frame lengths. The currently thought of method is to store them in a structure.
such as
     struct video_frame_info {
         int flags
         uint32_t frame_length;
     }
Add the structure to video_buffer as a pointer. Do you have any suggestions for us to modify this?
​

loicpoulain commented:
> @loicpoulain Your suggestions are very good. We have converted the API to a universal one, but we currently have a problem and hope that every fragment Frames all have frame lengths. Now, we have added them to video_buffer such as:
> 
> ```c
> struct video_buffer {
> 	/** pointer to driver specific data. */
> 	void *driver_data;
> 	/** pointer to the start of the buffer. */
> 	uint8_t *buffer;
> 	/** size of the buffer in bytes. */
> 	uint32_t size;
> 	/** number of bytes occupied by the valid data in the buffer. */
> 	uint32_t bytesused;
> 	/** time reference in milliseconds at which the last data byte was
> 	 * actually received for input endpoints or to be consumed for output
> 	 * endpoints.
> 	 */
> 	uint32_t timestamp;
> 	/** frame length for fragmented frames. */
> 	uint32_t frame_length;
> 	/** receiving status for fragmented frames. */
> 	enum video_frame_fragmented_status fragmented_status;
> };
> ```
> 
> Do you have any suggestions for us to modify this? ​

Ok, because AFAIU you need the total size before getting all the fragment, fair enough, but can you name it `bytesframe` to keep the same 'style' as `bytesused`, It would be the total frame length.

For the status, I would prefer a generic flags `uint32_t flags` (in case we want to extend with additional ones in the futur), with `VIDEO_BUF_FRAG` for indicating buffer is a frame fragment and `VIDEO_BUF_EOF` for indicating the end of frame (last fragment).

ArduCAM commented:
@loicpoulain 
Awesome.  We have modified the sample again followed your sugguestion. Please check it again.

loicpoulain reviewed (dismissed):
Great ! LGTM

ArduCAM commented:
@loicpoulain 
Thank you for your approval. Now waiting nashif and galak's review @nashif  @galak

kartben commented:
> @loicpoulain 
> Thank you for your approval. Now waiting nashif and galak's review @nashif  @galak 

@ArduCAM I understand (I really do!) that you are super eager to get this in but please refrain from tagging people every day or so :) pretty sure it just has the opposite effect of what you want :) Reviewers have limited cycles and unless a PR gets stuck for a couple weeks, I would recommend you don't do that. Thanks!

marcusfolkesson reviewed (approved):
LGTM

ArduCAM commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1026 @@
+		return -EIO;
+	}
+
+	return ret;
```
@demonday 
Hello, thank you for your comment on mega driver.  we just refer to the the ov2640 driver which has beed merged to the zephyr framework. I think the kernel drirver framework should process the issue you mentioned.

ArduCAM commented:
@loicpoulain 
Hello, we have done some modifications in the arducam_mega.c file 
1. Set up a dedicated workqueue for scheduling communication with the camera sensor over SPI. In order to prevent thread blocking and affect the work of other threads when using the same system work queue.
2. Change the framerate timing to use a zephyr timer, rather than rely on a delay, to make the timing more predictable.
Please check it again.

ArduCAM commented:
Conflict resolved again and submitted, what else is needed...

loicpoulain commented:
@ArduCAM looks quite ok, I understand you have a custom sample to show off all the camera features. But could you confirm or test that the driver also works with the simple video `capture` sample?

ArduCAM commented:
> @ArduCAM looks quite ok, I understand you have a custom sample to show off all the camera features. But could you confirm or test that the driver also works with the simple video `capture` sample?

Thanks for your reply. I will test it and reply you later.

ArduCAM commented:
> @ArduCAM looks quite ok, I understand you have a custom sample to show off all the camera features. But could you confirm or test that the driver also works with the simple video `capture` sample?

@loicpoulain 
Hello, I have some questions about the use of video capture. Freom the https://docs.zephyrproject.org/latest/samples/subsys/video/capture/README.html link，
I think the capture sample is used for  testing capture device ( such as camera) basic functions Using video API.
That's great. But the demo in the main branch is used on  mimxrt1064_evk as traget platform. If other platform, We have to modify the capture demo. Or I will prepare another capture demo and test API function?

ArduCAM commented:
> @ArduCAM looks quite ok, I understand you have a custom sample to show off all the camera features. But could you confirm or test that the driver also works with the simple video `capture` sample?

@loicpoulain Have  tested the capture using the folowing code
```
/*
 * Copyright (c) 2019 Linaro Limited
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#include <zephyr/kernel.h>
#include <zephyr/device.h>

#include <zephyr/drivers/video.h>

#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
#include <zephyr/logging/log.h>
LOG_MODULE_REGISTER(main);

#define VIDEO_DEV_SW "VIDEO_SW_GENERATOR"

int main(void)
{
	struct video_buffer *buffers[2], *vbuf;
	struct video_format fmt;
	struct video_caps caps;
	const struct device *video;
	unsigned int frame = 0;
	size_t bsize;
	size_t fsize;
	int i = 0;

	/* Default to software video pattern generator */
	video = device_get_binding(VIDEO_DEV_SW);
	if (video == NULL) {
		LOG_ERR("Video device %s not found", VIDEO_DEV_SW);
		return 0;
	}

	/* But would be better to use a real video device if any */
#if defined(CONFIG_VIDEO_MCUX_CSI)
	const struct device *const dev = DEVICE_DT_GET_ONE(nxp_imx_csi);

	if (!device_is_ready(dev)) {
		LOG_ERR("%s: device not ready.\n", dev->name);
		return 0;
	}

	video = dev;
#elif defined(CONFIG_VIDEO_ARDUCAM_MEGA)
	const struct device *const dev = DEVICE_DT_GET(DT_NODELABEL(arducam_mega0));

	if (!device_is_ready(dev)) {
		LOG_ERR("%s: device not ready.\n", dev->name);
		return 0;
	}
	video = dev;
	
#endif

	printk("- Device name: %s\n", video->name);

	/* Get capabilities */
	if (video_get_caps(video, VIDEO_EP_OUT, &caps)) {
		LOG_ERR("Unable to retrieve video capabilities");
		return 0;
	}

	printk("- Capabilities:\n");
	while (caps.format_caps[i].pixelformat) {
		const struct video_format_cap *fcap = &caps.format_caps[i];
		/* fourcc to string */
		printk("  %c%c%c%c width [%u; %u; %u] height [%u; %u; %u]\n",
		       (char)fcap->pixelformat,
		       (char)(fcap->pixelformat >> 8),
		       (char)(fcap->pixelformat >> 16),
		       (char)(fcap->pixelformat >> 24),
		       fcap->width_min, fcap->width_max, fcap->width_step,
		       fcap->height_min, fcap->height_max, fcap->height_step);
		i++;
	}

	/* Get default/native format */
	if (video_get_format(video, VIDEO_EP_OUT, &fmt)) {
		LOG_ERR("Unable to retrieve video format");
		return 0;
	}
	fmt.width = 320,
	fmt.height = 240,
	fmt.pixelformat = VIDEO_PIX_FMT_JPEG;
	video_set_format(video, VIDEO_EP_OUT, &fmt);

	printk("- Default format: %c%c%c%c %ux%u\n", (char)fmt.pixelformat,
	       (char)(fmt.pixelformat >> 8),
	       (char)(fmt.pixelformat >> 16),
	       (char)(fmt.pixelformat >> 24),
	       fmt.width, fmt.height);

	/* Size to allocate for each buffer */
	bsize = 1024;

	/* Alloc video buffers and enqueue for capture */
	for (i = 0; i < ARRAY_SIZE(buffers); i++) {
		buffers[i] = video_buffer_alloc(bsize);
		if (buffers[i] == NULL) {
			LOG_ERR("Unable to alloc video buffer");
			return 0;
		}

		video_enqueue(video, VIDEO_EP_OUT, buffers[i]);
	}
	printk("Prepare Capture started\n");
	/* Start video capture */
	if (video_stream_start(video)) {
		LOG_ERR("Unable to start capture (interface)");
		return 0;
	}

	printk("Capture started\n");

	/* Grab video frames */
	while (1) {
		int err;
		enum video_frame_fragmented_status f_status;
		err = video_dequeue(video, VIDEO_EP_OUT, &vbuf, K_FOREVER);
		if (err) {
			LOG_ERR("Unable to dequeue video buf");
			return 0;
		}
		 f_status = vbuf->flags;
	
		if (f_status == VIDEO_BUF_EOF){
			printk(" Got frame %u!;  size: %u; timestamp %u ms\n",
		       frame++, vbuf->bytesframe, vbuf->timestamp);
		}

		err = video_enqueue(video, VIDEO_EP_OUT, vbuf);
		if (err) {
			LOG_ERR("Unable to requeue video buf");
			return 0;
		}
	}
}
```
![image](https://github.com/zephyrproject-rtos/zephyr/assets/11016968/9ff1fb70-2bcd-4ff0-8eba-8e8e32e81eac)

ArduCAM commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1048 @@
+	arducam_mega_fifo_read(drv_data->dev, vbuf);
+
+	if (drv_data->fifo_length == 0) {
+		vbuf->flags = VIDEO_BUF_EOF;
```
Have marked resolved

danieldegrasse commented:
Looking at this API, I'm unclear how the consumer determines what size of framebuffer they should provide to the video device. In the case of the arducam, is the size essentially arbitrary? IE the application could provided 10 1KiB byte buffers, or 1 10KiB buffer, and the camera would work either way?

The other use case I'd like to try to cover with a "partial frame" API is one where the camera driver "slices" the frame into segments. For example, a 320x240 frame might be divided into 8 320x30 pixel buffers, which would allow a low memory device to stream QVGA frames in chunks. The issue here is how to tell the application what size of framebuffer is required, since in this case the size of each framebuffer is fixed.

In #72827, the approach I used was to provide a `vbufs_per_frame` field in the `video_caps` structure: https://github.com/zephyrproject-rtos/zephyr/pull/72827/files#diff-70c4de6949b97c13ac893b4bfb16fbcc433593fbfa96ec8947bb5cb83c4d04bc. This way, the video consumer can determine how many vbufs are included in one frame. However, this does not account for the case described here, where the size of the vbuf provided appears to be flexible.

@loicpoulain do you have any thoughts here? Maybe we add a special value to `vbufs_per_frame`, like `VIDEO_VBUFS_VARIABLE`, to indicate the driver can accept variably sized video buffers?

ArduCAM commented:
> Looking at this API, I'm unclear how the consumer determines what size of framebuffer they should provide to the video device. In the case of the arducam, is the size essentially arbitrary? IE the application could provided 10 1KiB byte buffers, or 1 10KiB buffer, and the camera would work either way?
> 
> The other use case I'd like to try to cover with a "partial frame" API is one where the camera driver "slices" the frame into segments. For example, a 320x240 frame might be divided into 8 320x30 pixel buffers, which would allow a low memory device to stream QVGA frames in chunks. The issue here is how to tell the application what size of framebuffer is required, since in this case the size of each framebuffer is fixed.
> 
> In #72827, the approach I used was to provide a `vbufs_per_frame` field in the `video_caps` structure: https://github.com/zephyrproject-rtos/zephyr/pull/72827/files#diff-70c4de6949b97c13ac893b4bfb16fbcc433593fbfa96ec8947bb5cb83c4d04bc. This way, the video consumer can determine how many vbufs are included in one frame. However, this does not account for the case described here, where the size of the vbuf provided appears to be flexible.
> 
> @loicpoulain do you have any thoughts here? Maybe we add a special value to `vbufs_per_frame`, like `VIDEO_VBUFS_VARIABLE`, to indicate the driver can accept variably sized video buffers?

Yes，For arducam spi camera, the the size of each  segments is flexible. Driver gets each frame size and send to user space through multi segments, due to some platform's memory limitation. For compressed data such as jpeg format,  the frame size is flexible and not stable. so for the last vbuf，it will contains some supplementary data, if the driver can accept variably sized video buffers, i think it‘s great.

danieldegrasse commented:
@loicpoulain and @ArduCAM, I've updated #72827 with commits to demonstrate how I think we might merge these two methods of handling partial video frames. Essentially, a driver would set the `VIDEO_CAP_VARIABLE_VBUFS` flag to indicate that it did not require a fixed video buffer size. Otherwise, the application should provide video buffer frames equal in size to `pitch * height / vbufs_per_frame`. The video driver will set the VIDEO_BUF_EOF flag on the last buffer in a frame (for drivers that use full frames, every buffer will have this flag set).

josuah commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1048 @@
+
+	ret |= arducam_mega_write_reg(&cfg->bus, CAM_REG_SHARPNESS_CONTROL, level);
+
+	if (ret == -1) {
```
Given the `-errno` being returned for errors, I suppose this would need to be converted to `ret != 0` or `ret < 0` (to support cases when ret > 0 to indicate status).

josuah commented on drivers/video/arducam_mega.c:
```diff
@@ -0,0 +1,1048 @@
+			2592, 1936, VIDEO_PIX_FMT_YUYV);
+		support_resolution[8] = MEGA_RESOLUTION_WQXGA2;
+		break;
+		drv_data->info = &mega_infos[0];
+		break;
```
This ooks like missing either a `case` either need to be removed.

josuah reviewed (changes requested):
Thank you for adding Zephyr support for this camera!

josuah commented:
It sounds like it is possible to merge this PR without introducing the extra capability immediately:

- The flags are information about what happened: If a frame did not fit the buffer, it tells so in the `vbuf->flags`.
- The `caps()` extensions tell how the device will behave: It will help the user sizing the buffers correctly.

But [it looks](https://github.com/zephyrproject-rtos/zephyr/pull/66994#issuecomment-2144098005) like Arducam does not need any `caps()` extension, so can be merged before they get introduced.

- If a buffer is too small, Arducam Mega will fragment the data with `VIDEO_BUF_FRAG`.
- If a buffer is too long, Arducam Mega will fill it partially with `VIDEO_BUF_EOF` on each frame.

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -96,6 +96,16 @@ struct video_caps {
+ *
+ * Indicates the receiving status of fragmented frames.
+ */
+enum video_frame_fragmented_status {
+	VIDEO_BUF_FRAG,
+	VIDEO_BUF_EOF,
+};
```
```suggestion
#define VIDEO_BUF_FRAG BIT(0)
#define VIDEO_BUF_EOF  BIT(1)
```

I think this is what @loicpoulain suggested [here](https://github.com/zephyrproject-rtos/zephyr/pull/66994#issuecomment-1905458088), but I could be wrong!

danieldegrasse commented:
> @ArduCAM wondering if you could rebase and address latest @josuah comments, I would like to move forward on this.

Could you also take a look at the API changes in #72827 and see if those API additions would handle this camera?

kartben commented:
@ArduCAM -- I know this has been open for a really long time, but as a lot has happened around video recently, I feel like addressing the remaining comments should help us get this really close to be mergeable. Thanks!
